CMakeLists.txt

```diff
@@ -179,24 +179,70 @@ endif ()
 
 # Reasonable default. CheckCompiler.m4 checks thoroughly for HP-UX's sake.
 set (MONGOC_SOCKET_ARG2 "struct sockaddr")
+set (MONGOC_SOCKET_ARG3 "socklen_t")
+
+if (NOT WIN32)
+
+set(MONGOC_ACCEPT_ARG2_LIST "struct sockaddr" "void")
+set(MONGOC_ACCEPT_ARG3_LIST "socklen_t" "size_t" "int")
+
+list(LENGTH MONGOC_ACCEPT_ARG2_LIST MONGOC_ACCEPT_ARG2_LIST_LEN)
+math(EXPR MONGOC_ACCEPT_ARG2_LIST_LEN_R "${MONGOC_ACCEPT_ARG2_LIST_LEN} - 1")
+list(LENGTH MONGOC_ACCEPT_ARG3_LIST MONGOC_ACCEPT_ARG3_LIST_LEN)
+math(EXPR MONGOC_ACCEPT_ARG3_LIST_LEN_R "${MONGOC_ACCEPT_ARG3_LIST_LEN} - 1")
+
+foreach(MONGOC_ACCEPT_ARG2_L RANGE ${MONGOC_ACCEPT_ARG2_LIST_LEN_R})
+foreach(MONGOC_ACCEPT_ARG3_L RANGE ${MONGOC_ACCEPT_ARG3_LIST_LEN_R})
+
+list(GET MONGOC_ACCEPT_ARG2_LIST ${MONGOC_ACCEPT_ARG2_L} MONGOC_ACCEPT_ARG2_VAL)
+list(GET MONGOC_ACCEPT_ARG3_LIST ${MONGOC_ACCEPT_ARG3_L} MONGOC_ACCEPT_ARG3_VAL)
+
+FILE(WRITE accept_test.cxx
+"#include <sys/types.h>
+ #include <sys/socket.h>
+
+ int main ()
+ {
+   int a = 0;
+   ${MONGOC_ACCEPT_ARG2_VAL} *b = 0;
+   ${MONGOC_ACCEPT_ARG3_VAL} *c = 0;
+   accept (a, b, c);
+   return 0;
+}
+")
+  TRY_COMPILE(RES ${CMAKE_CURRENT_BINARY_DIR}
+  ${CMAKE_CURRENT_SOURCE_DIR}/accept_test.cxx CMAKE_FLAGS
+  "-DCMAKE_CXX_LINK_EXECUTABLE='echo not linking now...'" OUTPUT_VARIABLE
+  LOG2)
```

Comment:
For the Autotools version it was important to pass "-Werror" to the compiler when testing each combination of parameter types, in order to ensure that we had found the *exact* right types, rather than a combination of types that the compiler would accept with a warning:

https://github.com/mongodb/mongo-c-driver/blob/master/build/autotools/m4/ax_prototype.m4#L220

Do you need to pass "-Werror" to clang and gcc here, and perhaps "/WX" if (MSVC)?